base/linalg/diagonal.jl

```diff
@@ -185,20 +185,23 @@ function A_mul_B!(D::Diagonal, B::UnitUpperTriangular)
     UpperTriangular(B.data)
 end
 
+Ac_mul_B(D::Diagonal, B::Diagonal) = Diagonal(conj(D.diag) .* B.diag)
 Ac_mul_B(A::AbstractTriangular, D::Diagonal) = A_mul_B!(ctranspose(A), D)
 function Ac_mul_B(A::AbstractMatrix, D::Diagonal)
     Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
     ctranspose!(Ac, A)
     A_mul_B!(Ac, D)
 end
 
+At_mul_B(D::Diagonal, B::Diagonal) = Diagonal(D.diag .* B.diag)
```

Comment:
IIRC `transpose` is recursive presently, requiring something along the lines of `Diagonal(transpose.(D.diag) .* B.diag)`?